packages/jest/src/generators/init/init.ts

```diff
@@ -34,74 +35,89 @@ const schemaDefaults = {
-
-    const ext = extname(rootJestConfig);
-    const newJestConfigPath = defaultProject
-      ? `jest.${defaultProject}.config${ext}`
-      : `jest.project-config${ext}`;
-
-    tree.rename(rootJestConfig, newJestConfigPath);
-
-    if (defaultProject) {
-      const projectConfig = readProjectConfiguration(tree, defaultProject);
-
-      // TODO(caleb): do we care about a custom target name?
-      if (projectConfig?.targets?.['test']?.options?.jestConfig) {
-        projectConfig.targets['test'].options.jestConfig = newJestConfigPath;
-        updateProjectConfiguration(tree, defaultProject, projectConfig);
+    rootJestPath &&
+    // TODO: (meeroslav) is this condition enough?
+    !tree.read(rootJestPath, 'utf-8').includes('getJestProjects()');
+
+  if (isProjectConfig) {
+    // moving from root project config to monorepo-style config
+    const projects = readWorkspace(tree).projects;
+    const projectNames = Object.keys(projects);
+    const rootProject = projectNames.find(
+      (projecName) => projects[projecName].root === '.'
+    );
+    // root project might have been removed,
+    // if it's missing there's nothing to migrate
+    if (rootProject) {
+      const jestTarget = Object.values(
+        projects[rootProject].targets || {}
+      ).find((t) => t.executor === '@nrwl/jest:jest');
+      // if root project doesn't have jest target, there's nothing to migrate
+      if (jestTarget) {
+        const pathSegments = rootJestPath
+          .split(FILE_EXTENSION_REGEX)
+          .filter(Boolean);
+        // insert root app name before the extension e.g. `jest.config.js` => `jest.config.myapp.js`
+        const rootProjJestPath = pathSegments.join(`.${rootProject}`);
+        tree.rename(rootJestPath, rootProjJestPath);
+        jestTarget.options.jestConfig = rootProjJestPath;
+        updateProjectConfiguration(tree, rootProject, projects[rootProject]);
       }
-    } else {
-      console.warn(stripIndents`Could not find the default project project.json to update the test target options.
-Manually update the 'jestConfig' path to point to ${newJestConfigPath}`);
     }
-  }
 
-  // if the root ts config already exists then don't make a js one or vice versa
-  if (!tree.exists('jest.config.ts') && !tree.exists('jest.config.js')) {
-    const contents = options.js
-      ? stripIndents`
-      const { getJestProjects } = require('@nrwl/jest');
-
-      module.exports = {
-        projects: getJestProjects()
-      };`
-      : stripIndents`
-      import { getJestProjects } from '@nrwl/jest';
-
-      export default {
-       projects: getJestProjects()
-      };`;
-    tree.write(`jest.config.${options.js ? 'js' : 'ts'}`, contents);
+    // original root file was renamed just now OR
+    // root project doesn't have target with jest executor OR
+    // root project was not found
+    // => generate new global config
+    generateGlobalConfig(tree, options.js);
```

Comment:
Whatever the case, if we started with `isProjectConfig` we need to replace it with proper global config.